extensions/ql-vscode/src/common/vscode/abstract-webview-view-provider.ts

```diff
@@ -0,0 +1,87 @@
+import * as vscode from "vscode";
+import { Uri, WebviewViewProvider } from "vscode";
+import { WebviewKind, WebviewMessage, getHtmlForWebview } from "./webview-html";
+import { Disposable } from "../disposable-object";
+import { App } from "../app";
+
+export abstract class AbstractWebviewViewProvider<
+  ToMessage extends WebviewMessage,
+  FromMessage extends WebviewMessage,
+> implements WebviewViewProvider
+{
+  protected webviewView: vscode.WebviewView | undefined = undefined;
+  private disposables: Disposable[] = [];
+
+  constructor(
+    private readonly app: App,
+    private readonly webviewKind: WebviewKind,
+  ) {}
+
+  /**
+   * This is called when a view first becomes visible. This may happen when the view is
+   * first loaded or when the user hides and then shows a view again.
+   */
+  public resolveWebviewView(
+    webviewView: vscode.WebviewView,
+    _context: vscode.WebviewViewResolveContext,
+    _token: vscode.CancellationToken,
+  ) {
+    webviewView.webview.options = {
+      enableScripts: true,
+      localResourceRoots: [Uri.file(this.app.extensionPath)],
+    };
+
+    const html = getHtmlForWebview(
+      this.app,
+      webviewView.webview,
+      this.webviewKind,
+      {
+        allowInlineStyles: true,
+        allowWasmEval: false,
+      },
+    );
+
+    webviewView.webview.html = html;
+
+    this.webviewView = webviewView;
+
+    webviewView.webview.onDidReceiveMessage(async (msg) => this.onMessage(msg));
+    webviewView.onDidDispose(() => this.dispose());
+
+    this.onWebViewLoaded();
+  }
+
+  protected get isShowingView() {
+    return this.webviewView?.visible ?? false;
+  }
+
+  protected async postMessage(msg: ToMessage): Promise<void> {
+    await this.webviewView?.webview.postMessage(msg);
+  }
+
+  protected dispose() {
+    while (this.disposables.length > 0) {
+      const disposable = this.disposables.pop()!;
+      disposable.dispose();
+    }
+
+    this.webviewView = undefined;
+  }
+
+  protected push<T extends Disposable>(obj: T): T {
+    if (obj !== undefined) {
+      this.disposables.push(obj);
+    }
+    return obj;
+  }
+
+  protected abstract onMessage(msg: FromMessage): Promise<void>;
+
+  /**
+   * This is called when a view first becomes visible. This may happen when the view is
+   * first loaded or when the user hides and then shows a view again.
+   */
+  protected onWebViewLoaded(): void {
```

Comment:
Additional question which I just encountered: This is called when the `viewLoaded` message is received in the `AbstractWebview`, but this is called after the panel is created in this class, without waiting for the webview to actually be loaded. That means you can't do `postMessage` in `onWebViewLoaded` because the React code to receive the messages may not have initialized yet.

There should probably be a requirement to handle the `viewLoaded` message in `onMessage` and call this method. If it's not handled, you will also get `ExhaustivityCheckingError` errors if you use `default: assertNever(msg)`.

Reply:
That also works, but I think we should still have the `onWebViewLoaded` and the requirement to handle `viewLoaded` as well. The view code (in `src/view/webview.tsx`) is shared by the `AbstractWebview` and this class, so we would still be receiving the `viewLoaded` message. If we don't handle the message, it may lead to surprising outcomes if you do `default: assertNever(msg)` because you would still be receiving the `viewLoaded` message.

Reply:
Good point! Updated.